src/FSharpPlus/Data/NonEmptyMap.fs

```diff
@@ -0,0 +1,266 @@
+    member x.TryFind(key) = x.Value.TryFind(key)
+    member x.TryGetValue(key, [<Out>] value:byref<'Value>) =
+      x.Value.TryGetValue(key, &value)
+
+    interface IReadOnlyDictionary<'Key, 'Value> with
+      member x.Item with get(key) = x.[key]
+      member x.Keys = (x.Value :> IReadOnlyDictionary<_, _>).Keys
+      member x.TryGetValue(key, value: byref<'Value>) = x.TryGetValue(key, &value)
+      member x.Values = (x.Value :> IReadOnlyDictionary<_, _>).Values
+      member x.ContainsKey key = x.ContainsKey key
+
+/// Basic operations on NonEmptyMap
+[<RequireQualifiedAccess>]
+module NonEmptyMap =
+    /// <summary>Builds a non empty map.</summary>
+    let create (k, v) (rest: ('k * 'v) seq) : NonEmptyMap<_, _> =
+      { Value = Map.ofSeq rest |> Map.add k v }
+    /// <summary>Builds a non empty map with a single element.</summary>
+    let singleton key value : NonEmptyMap<_, _> = { Value = Map.ofList [key, value] }
+
+    /// <summary>Returns a new map with the binding added to the given map.
+    /// If a binding with the given key already exists in the input map, the existing binding is replaced by the new binding in the result map.</summary>
+    /// <param name="key">The input key.</param>
+    /// <param name="value">The input value.</param>
+    /// <param name="table">The input map.</param>
+    /// <returns>The resulting map.</returns>
+    let add key value (table: NonEmptyMap<_, _>) = { Value = Map.add key value table.Value }
+
+    /// <summary>Builds a list from the given non empty map.</summary>
+    let toList  ({ Value = v }: NonEmptyMap<_, _>) = Map.toList v
+    /// <summary>Builds a sequence from the given non empty map.</summary>
+    let toSeq   ({ Value = v }: NonEmptyMap<_, _>) = Map.toSeq v
+    /// <summary>Builds an array from the given non empty map.</summary>
+    let toArray ({ Value = v }: NonEmptyMap<_, _>) = Map.toArray v
+    /// <summary>Builds a map from the given non empty map.</summary>
+    let toMap ({ Value = v }: NonEmptyMap<_, _>) = v
+
+    /// <summary>Builds a non empty map from the given array.</summary>
+    /// <param name="array">The input array.</param>
+    /// <returns>Non empty map containing the elements of the array.</returns>
+    /// <exception cref="System.ArgumentException">Thrown when the input array is empty.</exception>
+    /// <remarks>Throws exception for empty array</remarks>
+    let ofArray (array : _ array) =
+        match array |> Array.toList with
+        | []    -> invalidArg "array" "The input array was empty."
+        | x::xs -> create x xs
+    /// <summary>Builds a non empty map from the given list.</summary>
+    /// <param name="list">The input list.</param>
+    /// <returns>Non empty map containing the elements of the list.</returns>
+    /// <exception cref="System.ArgumentException">Thrown when the input list is empty.</exception>
+    /// <remarks>Throws exception for empty list</remarks>
+    let ofList (list : _ list) =
```

Comment:
It would be nice to have `ofNonEmptyList` as it would be a safe operation.
Yet better, `ofNonEmptySeq` if we had it ;)